metals/src/main/scala/scala/meta/internal/parsing/ClassFinder.scala

```diff
@@ -82,4 +83,136 @@ class ClassFinder(trees: Trees) {
+      newExtension: String,
+      checkInner: Boolean
+  ): Option[String] =
+    for {
+      tree <- trees.get(path)
+      val metaPos = pos.toMeta(tree.pos.input)
+      val fileWithoutExtension = path.filename.stripSuffix(".scala")
+      val TraverseResult(pkg, outer, dfnOpt) = traverseTree(
+        tree,
+        metaPos,
+        checkInner
+      )
+      dfn <- dfnOpt
+    } yield {
+      val filename = {
+        val basename = dfn match {
+          case Definition.Cls(name) => name
+          case Definition.Dfn => s"$fileWithoutExtension$$package"
+        }
+        s"$outer$basename.$newExtension"
+      }
+      (pkg ::: List(filename)).mkString("/")
+    }
+
+  private def traverseTree(
+      tree: Tree,
+      pos: Position,
+      checkInner: Boolean
+  ): TraverseResult = {
+    def extractPkgName(t: Term): List[String] = {
+      @tailrec
+      def loop(t: Term, acc: List[String]): List[String] =
+        t match {
+          case Term.Name(name) => (name :: acc)
+          case Term.Select(term, name) => loop(term, name.value :: acc)
+          case _ => acc
+        }
+      loop(t, Nil)
+    }
+
+    def findStat(
+        stats: List[Stat],
+        ifSome: TraverseResult,
+        ifNone: TraverseResult
+    ) =
+      stats
+        .find(s => s.pos.start <= pos.start && pos.start <= s.pos.end)
+        .orElse(stats.headOption) match {
+        case Some(value) => loop(value, ifSome)
+        case None => ifNone
+      }
+
+    @tailrec
+    def loop(tree: Tree, acc: TraverseResult): TraverseResult = {
+      tree match {
+        case Source(List(pkg: m.Pkg)) => loop(pkg, acc)
+        case Source(stats) => findStat(stats, acc, acc)
+
+        case Pkg(select, (p: Pkg) :: _) =>
+          val pkg = extractPkgName(select)
+          loop(p, acc.extendPackage(pkg))
+
+        case Pkg(select, stats) =>
+          val pkg = extractPkgName(select)
+          val newAcc = acc.extendPackage(pkg)
+          findStat(stats, newAcc, newAcc)
+
+        case obj: Defn.Object =>
+          if (!checkInner || obj.templ.stats.isEmpty)
+            acc.withDef(Definition.Cls(obj.name.value))
+          else
+            findStat(obj.templ.stats, acc.extendOuter(obj.name.value), acc)
+
+        case cls: Defn.Class =>
+          if (!checkInner || cls.templ.stats.isEmpty)
+            acc.withDef(Definition.Cls(cls.name.value))
+          else
+            findStat(cls.templ.stats, acc.extendOuter(cls.name.value), acc)
+
+        case trt: Defn.Trait =>
+          if (!checkInner || trt.templ.stats.isEmpty)
+            acc.withDef(Definition.Cls(trt.name.value))
+          else
+            findStat(trt.templ.stats, acc.extendOuter(trt.name.value), acc)
+
+        case _: Defn.Def =>
+          acc.withDef(Definition.Dfn)
+        case _ => acc
+      }
+    }
+
+    loop(tree, TraverseResult.empty)
+  }
+
+}
+
+private sealed trait Definition
+private object Definition {
+  case class Cls(name: String) extends Definition
```

Comment:
Why add the whole class hierarchy here instead of returning a string?

Reply:
It's leftover and it's unnecessary.